pkg/solana/chainwriter/lookups.go

```diff
@@ -210,63 +210,103 @@ func decodeBorshIntoType(data []byte, typ reflect.Type) (interface{}, error) {
-	var seedBytes [][]byte
+// The return type is [][][]byte, where each element of the outer slice is
+// one combination of seeds. This handles the case where one seed can resolve
+// to multiple addresses, multiplying the combinations accordingly.
+func getSeedBytes(
+	ctx context.Context,
+	lookup PDALookups,
+	args any,
+	derivedTableMap map[string]map[string][]*solana.AccountMeta,
+	reader client.Reader,
+) ([][][]byte, error) {
+	allCombinations := [][][]byte{
+		{},
+	}
 
+	// For each seed in the definition, expand the current list of combinations
+	// by all possible values for this seed.
 	for _, seed := range lookup.Seeds {
+		expansions := make([][]byte, 0)
 		if seed.Static != nil {
-			seedBytes = append(seedBytes, seed.Static)
+			expansions = append(expansions, seed.Static)
 		}
 		if seed.Dynamic != nil {
 			dynamicSeed := seed.Dynamic
 			if lookupSeed, ok := dynamicSeed.(AccountLookup); ok {
 				// Get value from a location (This doens't have to be an address, it can be any value)
 				bytes, err := GetValuesAtLocation(args, lookupSeed.Location)
 				if err != nil {
-					return nil, fmt.Errorf("error getting address seed: %w", err)
+					return nil, fmt.Errorf("error getting address seed for location %q: %w", lookupSeed.Location, err)
 				}
-				// validate seed length
+				// append each byte array to the expansions
 				for _, b := range bytes {
+					// validate seed length
 					if len(b) > solana.MaxSeedLength {
 						return nil, fmt.Errorf("seed byte array exceeds maximum length of %d: got %d bytes", solana.MaxSeedLength, len(b))
 					}
-					seedBytes = append(seedBytes, b)
+					expansions = append(expansions, b)
 				}
 			} else {
 				// Get address seeds from the lookup
 				seedAddresses, err := GetAddresses(ctx, args, []Lookup{dynamicSeed}, derivedTableMap, reader)
 				if err != nil {
 					return nil, fmt.Errorf("error getting address seed: %w", err)
 				}
-
-				// Add each address seed as bytes
-				for _, address := range seedAddresses {
-					seedBytes = append(seedBytes, address.PublicKey.Bytes())
+				// Add each address seed to the expansions
+				for _, addrMeta := range seedAddresses {
+					b := addrMeta.PublicKey.Bytes()
+					if len(b) > solana.MaxSeedLength {
+						return nil, fmt.Errorf("seed byte array exceeds maximum length of %d: got %d bytes", solana.MaxSeedLength, len(b))
+					}
+					expansions = append(expansions, b)
 				}
 			}
 		}
+
+		// expansions is the list of possible seed bytes for this single seed lookup.
+		// Multiply the existing combinations in allCombinations by each item in expansions.
+		newCombinations := make([][][]byte, 0, len(allCombinations)*len(expansions))
+		for _, existingCombo := range allCombinations {
+			for _, expandedSeed := range expansions {
+				comboCopy := make([][]byte, len(existingCombo))
```

Comment:
nit: If we know we're adding an element on top of this, should we initialize this with +1 length? Then instead of append below we can just set the new element through the index.

Reply:
Good suggestion, fixed!